extensions/amp-analytics/0.1/amp-analytics.js

```diff
@@ -331,12 +335,63 @@ export class AmpAnalytics extends AMP.BaseElement {
       request = addParamsToUrl(request, this.config_['extraUrlParams']);
     }
 
-    this.config_['vars']['requestCount']++;
 
+    return this.isSampledIn_(trigger).then(result => {
+      if (!result) {
+        return Promise.resolve('');
+      }
+
+      this.config_['vars']['requestCount']++;
+      request = this.expandTemplate_(request, trigger);
+
+      // For consistency with amp-pixel we also expand any url replacements.
+      return urlReplacementsFor(this.getWin()).expand(request).then(request => {
+        this.sendRequest_(request, trigger);
+        return request;
+      });
+    });
+  }
+
+  /**
+   * @param {!JSONObject} trigger The config to use to determine sampling.
+   * @return {!Promise.<boolean>} Whether the request should be sampled in or
+   * not based on sampleSpec.
+   * @private
+   */
+  isSampledIn_(trigger) {
+    const spec = trigger['sampleSpec'];
+    const resolve = Promise.resolve(true);
+    if (!spec) {
+      return resolve;
+    }
+    const threshold = spec['threshold'];
+    if (!spec['sampleOn'] ||
+        Number.isNaN(parseFloat(threshold)) || !Number.isFinite(threshold)) {
+      console./*OK*/error(this.getName_(), 'Invalid sampling spec.');
+      return resolve;
+    }
+    const key = this.expandTemplate_(spec['sampleOn'], trigger);
```

Comment:
On Mon, Apr 11, 2016 at 1:51 PM, Malte Ubl <notifications@github.com> wrote:

> In extensions/amp-analytics/0.1/amp-analytics.js
> <https://github.com/ampproject/amphtml/pull/2368#discussion_r59278956>:
>
> > +   * not based on sampleSpec.
> > +   * @private
> > +   */
> > +  isSampledIn_(trigger) {
> > +    const spec = trigger['sampleSpec'];
> > +    const resolve = Promise.resolve(true);
> > +    if (!spec) {
> > +      return resolve;
> > +    }
> > +    const threshold = spec['threshold'];
> > +    if (!spec['sampleOn'] ||
> > +        Number.isNaN(parseFloat(threshold)) || !Number.isFinite(threshold)) {
> > +      console./*OK*/error(this.getName_(), 'Invalid sampling spec.');
> > +      return resolve;
> > +    }
> > +    const key = this.expandTemplate_(spec['sampleOn'], trigger);
>
> What type of return value would this.expandTemplate_(spec['sampleOn'],
> trigger) typically give. You are using requestCount in the test. Is that
> a typical usage?
>

GA uses client id to sample hits. Some people might want random too.

> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/ampproject/amphtml/pull/2368/files/ddb55ee8e687cbb11c8fe886147cf218cd3d4e5e#r59278956>
>